csrc/rope.cu

```diff
@@ -259,3 +259,98 @@ void apply_llama31_rope_pos_ids(at::Tensor q, at::Tensor k, at::Tensor q_rope, a
+  CHECK_LAST_DIM_CONTIGUOUS(k_nope_in);
+  CHECK_LAST_DIM_CONTIGUOUS(q_rope_out);
+  CHECK_LAST_DIM_CONTIGUOUS(k_rope_out);
+  CHECK_LAST_DIM_CONTIGUOUS(q_nope_out);
+  CHECK_LAST_DIM_CONTIGUOUS(k_nope_out);
+  CHECK_INPUT(cos_sin_cache);
+  CHECK_INPUT(pos_ids);
+
+  CHECK_EQ(q_rope_in.size(-1), 64);
+  CHECK_EQ(k_rope_in.size(-1), 64);
+  CHECK_EQ(q_nope_in.size(-1), 512);
+  CHECK_EQ(k_nope_in.size(-1), 512);
+  CHECK_EQ(q_rope_out.size(-1), 64);
+  CHECK_EQ(k_rope_out.size(-1), 64);
+  CHECK_EQ(q_nope_out.size(-1), 512);
+  CHECK_EQ(k_nope_out.size(-1), 512);
+  auto scalar_type_in = q_rope_in.scalar_type();
+  TORCH_CHECK(scalar_type_in == k_rope_in.scalar_type());
+  TORCH_CHECK(scalar_type_in == q_nope_in.scalar_type());
+  TORCH_CHECK(scalar_type_in == k_nope_in.scalar_type());
+  auto quant_type_out = q_rope_out.scalar_type();
+  TORCH_CHECK(quant_type_out == k_rope_out.scalar_type());
+  TORCH_CHECK(quant_type_out == q_nope_out.scalar_type());
+  TORCH_CHECK(quant_type_out == k_nope_out.scalar_type());
+
+  CHECK_DIM(3, q_rope_in);   // q_rope_in: (nnz, H_Q, 64)
+  CHECK_DIM(3, q_nope_in);   // q_nope_in: (nnz, H_Q, 512)
+  CHECK_DIM(2, k_rope_in);   // k_rope_in: (nnz, 64)
+  CHECK_DIM(2, k_nope_in);   // k_nope_in: (nnz, 512)
+  CHECK_DIM(3, q_rope_out);  // q_rope_out: (nnz, H_Q, 64)
+  CHECK_DIM(3, q_nope_out);  // q_nope_out: (nnz, H_Q, 512)
+  CHECK_DIM(2, k_rope_out);  // k_rope_out: (nnz, 64)
+  CHECK_DIM(2, k_nope_out);  // k_nope_out: (nnz, 512)
+  uint32_t nnz = q_rope_in.size(0);
+  CHECK_EQ(q_nope_in.size(0), nnz);
+  CHECK_EQ(k_nope_in.size(0), nnz);
+  CHECK_EQ(q_rope_out.size(0), nnz);
+  CHECK_EQ(k_rope_out.size(0), nnz);
+  CHECK_EQ(q_nope_out.size(0), nnz);
+  CHECK_EQ(k_nope_out.size(0), nnz);
+  uint32_t num_heads = q_rope_in.size(1);
+  CHECK_EQ(q_rope_in.size(1), num_heads);
+  CHECK_EQ(q_nope_in.size(1), num_heads);
+  CHECK_EQ(q_rope_out.size(1), num_heads);
+  CHECK_EQ(q_nope_out.size(1), num_heads);
+
+  const uint32_t q_rope_in_stride_n = q_rope_in.stride(0);
+  const uint32_t q_rope_in_stride_h = q_rope_in.stride(1);
+  const uint32_t q_nope_in_stride_n = q_nope_in.stride(0);
+  const uint32_t q_nope_in_stride_h = q_nope_in.stride(1);
+  const uint32_t q_rope_out_stride_n = q_rope_out.stride(0);
+  const uint32_t q_rope_out_stride_h = q_rope_out.stride(1);
+  const uint32_t q_nope_out_stride_n = q_nope_out.stride(0);
+  const uint32_t q_nope_out_stride_h = q_nope_out.stride(1);
+  const uint32_t k_rope_in_stride = k_rope_in.stride(0);
+  const uint32_t k_nope_in_stride = k_nope_in.stride(0);
+  const uint32_t k_rope_out_stride = k_rope_out.stride(0);
+  const uint32_t k_nope_out_stride = k_nope_out.stride(0);
+
+  const c10::cuda::OptionalCUDAGuard device_guard(q_rope_in.device());
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP16(scalar_type_in, c_type, [&] {
+    return DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP8(quant_type_out, c_quant_type, [&] {
```

Comment:
this means quantization must happen right, and if `quant_type_out` is fp16 this would not work.

Is it possible to make quant optional and fuse rope + q_nope weight absorb for fp16 and rope + q_nope weight absorb + quant for fp8. Main reason for this, is integration code will be cleaner as we will use this kernel for both dtypes and also we can avoid concat on fp16 case as well.